### PR TITLE
[BE] Prefer dash over underscore in command-line options

### PR DESCRIPTION
**Pull request by XuehaiPan**
Preferring dash over underscore in command-line options. Add `--command-arg-name` to the argument parser. The old arguments with underscores `--command_arg_name` are kept for backward compatibility.

Both dashes and underscores are used in the PyTorch codebase. Some argument parsers only have dashes or only have underscores in arguments. For example, the `torchrun` utility for distributed training only accepts underscore arguments (e.g., `--master_port`). The dashes are more common in other command-line tools. And it looks to be the default choice in the Python standard library:

`argparse.BooleanOptionalAction`: https://github.com/python/cpython/blob/4a9dff0e5adc91cbb1ed68c495dac64ccfe608bd/Lib/argparse.py#L893-L895

```python
class BooleanOptionalAction(Action):
    def __init__(...):
            if option_string.startswith('--'):
                option_string = '--no-' + option_string[2:]
                _option_strings.append(option_string)
```

It adds `--no-argname`, not `--no_argname`. Also typing `_` need to press the shift or the caps-lock key than `-`.

cc @EikanWang @jgong5 @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

**Review by ezyang — APPROVED** [review]
OK sure why not

**Review by seemethere — APPROVED** [review]
LGTM, thanks for considering the backwards compat case!

File: torch/distributed/run.py [hunk 1]
```diff
@@ -729,12 +744,12 @@ def config_from_args(args) -> Tuple[LaunchConfig, Union[Callable, str], List[str
         else:
             if args.module:
                 raise ValueError(
-                    "Don't use both the '--no_python' flag"
+                    "Don't use both the '--no-python' flag"
                     " and the '--module' flag at the same time."
                 )
             cmd = args.training_script
     if not use_env:
-        cmd_args.append(f"--local_rank={macros.local_rank}")
```
Comment (antony-frolov) [inline torch/distributed/run.py:737]: Hi! This change broke our code and probably not only ours. Can we still have an option to use `--local_rank` with an underscore?

Comment (ezyang) [reply torch/distributed/run.py:737]: sure, can you send us a patch

**Comment by chedatomasz** [comment]
+1 on the change from --local_rank to --local-rank breaking scripts. This is very much NOT backward compatible, and there doesn't seem to be an easy fix on the framework side.
Given how much care was taken deprecating torch.distributed.launch in favor of torchrun, this seems like a serious oversight, as the required user intervention is really similar. 
This is also VERY hard to debug, given how argparse mangles hyphens into underscores, and how the docs make no mention of the change (If you have the time, please add a disclaimer). For searchability, the error message is similar to `error: unrecognized arguments: --local-rank=0`.

>

**Comment by XuehaiPan** [comment]
Hi @chedatomasz, I'm sorry for the inconvenience caused by this change.

In `torchrun`, the dashed option `--local-rank={rank}` will passed to the argument parser in the user script. This is a breaking change if the user argument parser only accepts underscored option `--local_rank={rank}`. The fix is very simple. You can add `'--local-rank'` to your argument parser while keeping underscored `--local_rank` for backward compatibility.

```diff
parser.add_argument(
+   '--local-rank',
    '--local_rank',
    type=int,
    ...
)
```

The `--local-rank` option is the only option we add to the training script arguments:

https://github.com/pytorch/pytorch/blob/f0fb4b3897e9cac3b99ee5b9b2ecab255e9e2da3/torch/distributed/run.py#L762-L764

We can only add either `--local-rank={rank}` or `--local_rank={rank}` to `cmd_args`, not both. Documentation and guidelines would be very necessary.

------

> This is also VERY hard to debug, given how argparse mangles hyphens into underscores, and how the docs make no mention of the change (If you have the time, please add a disclaimer). For searchability, the error message is similar to `error: unrecognized arguments: --local-rank=0`.

I will submit a PR to document this.

**Comment by CauchyComplete** [comment]
Due to this change, Torch 2.0 is no longer perfectly backward compatible with Torch 1.13. While torchrun uses local-rank, legacy codes require local_rank. Therefore, running past DDP code with Torch 2 results in errors. It is very inconvenient.

**Pull request closed without merging**